### PR TITLE
[BPF] DSR optout CIDRs

### PR DESCRIPTION
**Pull request by tomastigera**
Setting BPFDSROptoutCIDRs to a list of CIDRs allows clients from these CIDRs to opt out from DSR when DSR is enabled. That allows to use DSR in the part of your network where it works and keep using standard nodeports in that part of the network where it is not possible to use DSR.

BPFDSROptoutCIDRs can overlaps with other CIDRs

For instance 0.0.0.0/0 would optout all clients from DSR. Those CIDRs
must be either disjoint with other CIDRs in the cluster or their
super-sets. That means, for instance, if an IP pool overlaps only
partially with the opt out CIDR, the optout may or may not be applied as
it is applied to the route blocks on each node, but does not split those
blocks.

The BPFDSROptoutCIDRs is meant to be used with external entities to the 
cluser. It has no effect on in-cluster entities since DSR does not 
have much meaning fot them anyway.

A useful case may be to apply this to external nodes, provider specific
ip like AKS healt check IPs etc.

## Todos

- [ ] Tests
- [ ] Documentation
- [ ] Release note

## Release Note

```release-note
ebpf: Setting BPFDSROptoutCIDRs to a list of CIDRs allows clients from these CIDRs to opt out from DSR when DSR is enabled.
```

## Reminder for the reviewer

Make sure that this PR has the correct labels and milestone set.

Every PR needs one `docs-*` label.

- `docs-pr-required`: This change requires a change to the documentation that has not been completed yet.
- `docs-completed`: This change has all necessary documentation completed.
- `docs-not-required`: This change has no user-facing impact and requires no docs.

Every PR needs one `release-note-*` label.

- `release-note-required`: This PR has user-facing changes. Most PRs should have this label.
- `release-note-not-required`: This PR has no user-facing changes.

Other optional labels:

- `cherry-pick-candidate`: This PR should be cherry-picked to an earlier release. For bug fixes only.
- `needs-operator-pr`: This PR is related to install and requires a corresponding change to the operator.

### CODE REVIEW

File: api/pkg/apis/projectcalico/v3/felixconfig.go
Comment (fasaxc) [inline api/pkg/apis/projectcalico/v3/felixconfig.go:387]: Was surprised to see the pointer to a slice (slices already have a nil value) but it seems like we do it elsewhere :thinking: . Maybe it allows for a node-specific felix config to override the global one to `nil`.  Not sure that will actually work with omitempty etc.

File: felix/bpf-gpl/globals.h
Comment (fasaxc) [inline felix/bpf-gpl/globals.h:32]:
```suggestion
	CALI_GLOBALS_NO_DSR_CIDRS		= 0x00000040,
```

File: felix/dataplane/linux/bpf_route_mgr.go
Comment (fasaxc) [inline felix/dataplane/linux/bpf_route_mgr.go:249]: `For this feature I think you need to check if the route is inside any of the `dsrOptoutCIDRs` For (bad) example, if the user sets `DSROptout=0.0.0.0/0` then all traffic should be excluded, even if it's also a node IP, or something from an IP pool.  The logic will put 0.0.0.0/0 in the routes map but the IP pool routes will be excluded.

I think we've got a trie implementation in the ip package, should be trivial to use the trie instead of the set and then the contains check will just do the right thing.

Comment (fasaxc) [reply felix/dataplane/linux/bpf_route_mgr.go:249]: Yep, `CIDRTrie.Covers` does the check

File: felix/bpf/ut/nat_test.go
Comment (fasaxc) [inline felix/bpf/ut/nat_test.go:1899]: Wonder if you could refactor a bit to make the between-packet set-up common and maybe add more comments to really spell out what you're doing.  I think that'd aid readability, especially for folks who aren't familiar with these tests
```
// Pretend that we're on node 1, set up NAT maps with a node port.
maps := mustSetUpMaps(WithNodePort(...)) 

// Simulate a packet arriving at node 1.  Expect the BFP program to encap the packet and redirect to node 2.
...

// Now pretend that we're on node 2...  Configure the NAT maps again.
...
// Feed it the packet from node 1, expect it to spot that it's a no-DSR source
// and record that in conntrack before forwarding to a local pod.

// Now pretend we've got the response packet form the pod at egress from the pod...

```

**Review by fasaxc — CHANGES_REQUESTED** [review]
Nice; just one potential issue

File: felix/dataplane/linux/bpf_route_mgr.go
Comment (tomastigera) [reply felix/dataplane/linux/bpf_route_mgr.go:249]: OK, excellent, I was pondering that and then decided that it is meant for external cidrs only. This will make it robust :+1:

File: api/pkg/apis/projectcalico/v3/felixconfig.go
Comment (tomastigera) [reply api/pkg/apis/projectcalico/v3/felixconfig.go:387]: yup me also surprised :man_shrugging:

**Comment by marvin-tigera** [comment]
OK, I will merge the pull request when it's ready, leave the commits as is when I merge it, and leave the branch after I've merged it.

**Review by fasaxc — APPROVED** [review]
LGTM!

**Pull request merged**
